src/sentry/api/endpoints/project_repo_path_parsing.py

```diff
@@ -0,0 +1,131 @@
+from sentry.api.bases.project import ProjectEndpoint
+from sentry.api.serializers.rest_framework.base import CamelSnakeSerializer
+from sentry.models import Integration, Repository
+from sentry.utils.compat import filter, map
+from sentry.web.decorators import transaction_start
+
+
+def find_roots(stack_path, source_path):
+    """
+        Returns a tuple containing the stack_root, and the source_root.
+        If there is no overlap, raise an exception since this should not happen
+    """
+    overlap_to_check = stack_path
+    stack_root = ""
+    while overlap_to_check:
+        # see if our path ends with the overlap we want
+        if source_path.endswith(overlap_to_check):
+            # determine the source root by removing the overlap
+            source_root = source_path.rpartition(overlap_to_check)[0]
+            return (stack_root, source_root)
+        # increase the stack root specificity
+        # while decreasing the overlap
+        stack_root += overlap_to_check[0]
+        overlap_to_check = overlap_to_check[1:]
+    # validate_source_url should have ensured the file names match
+    # so if we get here something went wrong and there is a bug
+    raise Exception("Could not common root from paths")
+
+
+class PathMappingSerializer(CamelSnakeSerializer):
+    stack_path = serializers.CharField()
+    source_url = serializers.URLField()
+
+    def __init__(self, *args, **kwargs):
+        super(PathMappingSerializer, self).__init__(*args, **kwargs)
+        self.integration = None
+        self.repo = None
+
+    @property
+    def providers(self):
+        # TODO: use feature flag in the future
+        providers = filter(lambda x: x.has_stacktrace_linking, list(integrations.all()))
+        return map(lambda x: x.key, providers)
+
+    @property
+    def org_id(self):
+        return self.context["organization_id"]
+
+    def validate_source_url(self, source_url):
+        # first check to see if we are even looking at the same file
+        stack_path = self.initial_data["stack_path"]
+        stack_file = stack_path.split("/")[-1]
+        source_file = source_url.split("/")[-1]
+
+        if stack_file != source_file:
+            raise serializers.ValidationError(
+                "Source code URL points to a different file than the stack trace"
+            )
+
+        def integration_match(integration):
+            return source_url.startswith(u"https://{}".format(integration.metadata["domain_name"]))
+
+        def repo_match(repo):
+            return source_url.startswith(repo.url)
+
+        # now find the matching integration
+        integrations = Integration.objects.filter(
+            organizations=self.org_id, provider__in=self.providers
+        )
+
+        matching_integrations = filter(integration_match, integrations)
+        if not matching_integrations:
+            raise serializers.ValidationError("Could not find integration")
+
+        self.integration = matching_integrations[0]
```

Comment:
do we know what number of organizations have multiple installations for GH or GitLab (not onprem GitLab)?

Reply:
@MeredithAnya Not sure but it shouldn't matter since we are filtering on based on the `domain_name`. There should be only one match there. We can throw an assertion error if there are multiple matches if you think that makes sense.

Reply:
@scefali ah okays yeah had to check just now that the `domain_name` includes both the domain and the org, cool